File: src/com/easy/SumOfLeftLeaves_404.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class SumOfLeftLeaves_404 {
    // Use recursive:
    // Time Complexity: O(n)
    // Space Complexity: O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 36.4 MB, less than 100.00%
    public int sumOfLeftLeaves(TreeNode root) {
        if (root == null) {
            return 0;
        }
        if(bIsLeaf(root.left)) {
            return root.left.val + sumOfLeftLeaves(root.right);
        }
        return (sumOfLeftLeaves(root.left) + sumOfLeftLeaves(root.right));
    }

    public boolean bIsLeaf(TreeNode node) {
        if (node == null) {
            return false;
        }
        return ((node.left == null) && (node.right == null));
    }
}

File: Golang_src/easy/sumOfLeftLeaves_404.go
package easy


// Use recursive:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.7 MB, less than 100.00%
func sumOfLeftLeaves(root *TreeNode) int {
	if root == nil {
		return 0
	}

	if checkLeaf(root.Left) {
		return root.Left.Val + sumOfLeftLeaves(root.Right)
	}

	return sumOfLeftLeaves(root.Left) + sumOfLeftLeaves(root.Right)
}

func checkLeaf(root *TreeNode) bool {
	if root == nil {
		return false
	}
	if root.Left == nil && root.Right == nil {
		return true
	}

	return false
}

// Use iterative:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.6 MB, less than 100.00%
func sumOfLeftLeaves2(root *TreeNode) int {
	if root == nil {
		return 0
	}

	queue := []*TreeNode{root}
	result := 0

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		if checkLeaf(node.Left) {
			result += node.Left.Val
		}
		if node.Left != nil {
			queue = append(queue, node.Left)
		}
		if node.Right != nil {
			queue = append(queue, node.Right)
		}
	}

	return result
}
